pkg/dns/utils/trylock.go

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ *
+ */
+
+package utils
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// TryLock is a lock supporting both `Lock` and `TryLock` by wrapping a weighted semaphore.
+type TryLock struct {
+	lock *semaphore.Weighted
+	ctx  context.Context
+}
+
+// NewTryLock creates a lock based on a weighted semaphore.
+func NewTryLock() *TryLock {
```

Comment:
optionally take  a context.

And pass the context from the controller to assure a deblock if the controller is stopped.

Reply:
optional context argument has been added